File: src/ingestion/chains/interspar.ts
/**
 * Interspar Chain Adapter
 *
 * Adapter for parsing Interspar retail chain price data files.
 * Interspar uses CSV format with semicolon delimiter and UTF-8 encoding.
 * Store resolution is based on filename.
 *
 * Interspar portal: https://www.spar.hr/usluge/cjenici
 * URL format: https://www.spar.hr/usluge/cjenici?date=YYYY-MM-DD
 * Download links: /cjenik/download?file=FILENAME
 */

import type { DiscoveredFile } from '../core/types'
import type { CsvColumnMapping } from '../parsers/csv'
import { BaseCsvAdapter } from './base'
import { CHAIN_CONFIGS } from './config'

/**
 * Interspar JSON API response structure.
 * The API endpoint returns a list of available price files.
 */
interface IntersparJsonResponse {
  files: Array<{
    name: string
    URL: string
    SHA: string
  }>
}

/**
 * Column mapping for Interspar CSV files.
 * Maps Interspar's column names to NormalizedRow fields.
 *
 * Based on actual Interspar CSV format:
 * naziv;šifra;marka;neto količina;jedinica mjere;MPC (EUR);cijena za jedinicu mjere (EUR);
 * MPC za vrijeme posebnog oblika prodaje (EUR);Najniža cijena u posljednjih 30 dana (EUR);
 * sidrena cijena na DATE. (EUR);barkod;kategorija proizvoda
 */
const INTERSPAR_COLUMN_MAPPING: CsvColumnMapping = {
  externalId: 'šifra',
  name: 'naziv',
  category: 'kategorija proizvoda',
  brand: 'marka',
  unit: 'jedinica mjere',
  unitQuantity: 'neto količina',
  price: 'MPC (EUR)',
  discountPrice: 'MPC za vrijeme posebnog oblika prodaje (EUR)',
  barcodes: 'barkod',
  // Croatian price transparency fields
  unitPrice: 'cijena za jedinicu mjere (EUR)',
  lowestPrice30d: 'Najniža cijena u posljednjih 30 dana (EUR)',
  anchorPrice: 'sidrena cijena na 2.5.2025. (EUR)',
}

/**
 * Alternative column mapping for Interspar CSV files.
 * Some Interspar exports may use abbreviated or different column names.
 */
const INTERSPAR_COLUMN_MAPPING_ALT: CsvColumnMapping = {
  externalId: 'Šifra',
  name: 'Naziv',
  category: 'Kategorija',
  brand: 'Marka',
  unit: 'Mjerna jedinica',
  unitQuantity: 'Količina',
  price: 'Cijena',
  discountPrice: 'Akcijska cijena',
  discountStart: 'Početak akcije',
  discountEnd: 'Kraj akcije',
  barcodes: 'Barkod',
  // Croatian price transparency fields
  unitPrice: 'Cijena za jedinicu mjere',
  lowestPrice30d: 'Najniža cijena u zadnjih 30 dana',
  anchorPrice: 'Sidrena cijena',
  unitPriceBaseQuantity: 'Količina za jedinicu mjere',
  unitPriceBaseUnit: 'Jedinica mjere za cijenu',
  anchorPriceAsOf: 'Datum sidrene cijene',
}

/**
 * Interspar chain adapter implementation.
 * Extends BaseCsvAdapter for common CSV parsing functionality.
 *
 * Supports single-date discovery via setDiscoveryDate() method.
 */
export class IntersparAdapter extends BaseCsvAdapter {
  /** Date to discover files for (YYYY-MM-DD format, set by CLI before discovery) */
  private discoveryDate: string | null = null

  constructor() {
    super({
      slug: 'interspar',
      name: 'Interspar',
      supportedTypes: ['csv'],
      chainConfig: CHAIN_CONFIGS.interspar,
      columnMapping: INTERSPAR_COLUMN_MAPPING,
      alternativeColumnMapping: INTERSPAR_COLUMN_MAPPING_ALT,
      filenamePrefixPatterns: [
        /^Interspar[_-]?/i,
        /^Spar[_-]?/i,
        /^cjenik[_-]?/i,
      ],
      rateLimitConfig: {
        requestsPerSecond: 2,
        maxRetries: 3,
      },
    })
  }

  /**
   * Set the date to use for discovery.
   * @param date - Date in YYYY-MM-DD format
   */
  setDiscoveryDate(date: string): void {
    this.discoveryDate = date
  }

  /**
   * Discover available price files from Interspar portal.
   *
   * Interspar provides a JSON API endpoint that returns all available files:
   * - API URL: /datoteke_cjenici/Cjenik{YYYYMMDD}.json
   * - Returns: { "files": [{ "name": "...", "URL": "...", "SHA": "..." }] }
   *
   * This is much more efficient than scraping the HTML page.
   *
   * @returns Array of discovered files for the specified date
   */
  async discover(): Promise<DiscoveredFile[]> {
    const discoveredFiles: DiscoveredFile[] = []

    // Use provided date or default to today
    const date = this.discoveryDate || new Date().toISOString().split('T')[0]

    // Convert date from YYYY-MM-DD to YYYYMMDD format for the API
    const dateForApi = date.replace(/-/g, '')

    // Construct the JSON API URL
    const apiUrl = `https://www.spar.hr/datoteke_cjenici/Cjenik${dateForApi}.json`
    console.log(`[DEBUG] Fetching Interspar JSON API: ${apiUrl}`)

    try {
      const response = await fetch(apiUrl, {
        headers: {
          'User-Agent': 'Mozilla/5.0 (compatible; PriceTracker/1.0)',
          'Accept': 'application/json',
        },
      })

      if (!response.ok) {
        console.error(`Failed to fetch Interspar JSON API: ${response.status} ${response.statusText}`)
        console.error(`  URL: ${apiUrl}`)
        return []
      }

      const data = await response.json() as IntersparJsonResponse

      if (!data.files || data.files.length === 0) {
        console.log(`[INFO] No files found in JSON response for date ${date}`)
        return []
      }

      console.log(`[DEBUG] Found ${data.files.length} file(s) in JSON response`)

      for (const file of data.files) {
        discoveredFiles.push({
          url: file.URL,
          filename: file.name,
          type: 'csv',
          size: null,
          lastModified: new Date(date),
          metadata: {
            source: 'interspar_json_api',
            discoveredAt: new Date().toISOString(),
            portalDate: date,
            sha: file.SHA,
          },
        })
      }
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error)
      console.error(`Error discovering Interspar files: ${errorMessage}`)
      console.error(`  URL: ${apiUrl}`)
    }

    return discoveredFiles
  }

  /**
   * Extract store identifier from Interspar filename.
   *
   * Interspar filenames may follow patterns like:
   * - INTERSPAR_STORE_ID_DATE.csv
   * - Cjenik_Interspar_LOCATION_DATE.csv
   *
   * @param filename - The CSV filename
   * @returns Store identifier string
   */
  protected override extractStoreIdentifierFromFilename(filename: string): string {
    // Remove file extension
    const baseName = filename.replace(/\.(csv|CSV)$/, '')

    // Try to extract store ID from various Interspar filename patterns
    // Pattern 1: Contains a 4-digit store code
    const storeCodeMatch = baseName.match(/[_-](\d{4})[_-]/)
    if (storeCodeMatch) {
      return storeCodeMatch[1]
    }

    // Pattern 2: Store location name after Interspar prefix
    const locationMatch = baseName.match(/^(?:Interspar|Spar)[_-]?(.+?)(?:[_-]\d{4}[_-]\d{2}[_-]\d{2})?$/i)
    if (locationMatch) {
      return locationMatch[1]
    }

    // Fall back to base class implementation
    return super.extractStoreIdentifierFromFilename(filename)
  }
}

/**
 * Create an Interspar adapter instance.
 */
export function createIntersparAdapter(): IntersparAdapter {
  return new IntersparAdapter()
}

File: services/price-service/internal/adapters/chains/interspar.go
package chains

import (
	"encoding/json"
	"fmt"
	"io"
	"regexp"
	"strings"
	"time"

	"github.com/kosarica/price-service/internal/adapters/base"
	"github.com/kosarica/price-service/internal/adapters/config"
	"github.com/kosarica/price-service/internal/parsers/csv"
	"github.com/kosarica/price-service/internal/types"
)

// intersparJsonFile represents a file entry from Interspar's JSON API
type intersparJsonFile struct {
	Name string `json:"name"`
	URL  string `json:"URL"`
	SHA  string `json:"SHA"`
}

// intersparJsonResponse represents the JSON API response structure
type intersparJsonResponse struct {
	Files []intersparJsonFile `json:"files"`
}

// intersparColumnMapping is the primary column mapping for Interspar CSV files
var intersparColumnMapping = csv.CsvColumnMapping{
	ExternalID:     types.StringPtr("šifra"),
	Name:           "naziv",
	Category:       types.StringPtr("kategorija proizvoda"),
	Brand:          types.StringPtr("marka"),
	Unit:           types.StringPtr("jedinica mjere"),
	UnitQuantity:   types.StringPtr("neto količina"),
	Price:          "MPC (EUR)",
	DiscountPrice:  types.StringPtr("MPC za vrijeme posebnog oblika prodaje (EUR)"),
	Barcodes:       types.StringPtr("barkod"),
	UnitPrice:      types.StringPtr("cijena za jedinicu mjere (EUR)"),
	LowestPrice30d: types.StringPtr("Najniža cijena u posljednjih 30 dana (EUR)"),
	AnchorPrice:    types.StringPtr("sidrena cijena na 2.5.2025. (EUR)"),
}

// intersparColumnMappingAlt is the alternative column mapping for Interspar CSV files
var intersparColumnMappingAlt = csv.CsvColumnMapping{
	ExternalID:          types.StringPtr("Šifra"),
	Name:                "Naziv",
	Category:            types.StringPtr("Kategorija"),
	Brand:               types.StringPtr("Marka"),
	Unit:                types.StringPtr("Mjerna jedinica"),
	UnitQuantity:        types.StringPtr("Količina"),
	Price:               "Cijena",
	DiscountPrice:       types.StringPtr("Akcijska cijena"),
	DiscountStart:       types.StringPtr("Početak akcije"),
	DiscountEnd:         types.StringPtr("Kraj akcije"),
	Barcodes:            types.StringPtr("Barkod"),
	UnitPrice:           types.StringPtr("Cijena za jedinicu mjere"),
	LowestPrice30d:      types.StringPtr("Najniža cijena u zadnjih 30 dana"),
	AnchorPrice:         types.StringPtr("Sidrena cijena"),
	UnitPriceBaseQuantity: types.StringPtr("Količina za jedinicu mjere"),
	UnitPriceBaseUnit:   types.StringPtr("Jedinica mjere za cijenu"),
	AnchorPriceAsOf:     types.StringPtr("Datum sidrene cijene"),
}

// IntersparAdapter is the chain adapter for Interspar retail chain
type IntersparAdapter struct {
	*base.BaseCsvAdapter
	discoveryDate string // Date filter for discovery (YYYY-MM-DD format)
}

// NewIntersparAdapter creates a new Interspar adapter
func NewIntersparAdapter() (*IntersparAdapter, error) {
	chainConfig := config.ChainConfigs[config.ChainInterspar]

	adapterConfig := base.CsvAdapterConfig{
		BaseAdapterConfig: base.BaseAdapterConfig{
			Slug:           string(config.ChainInterspar),
			Name:           chainConfig.Name,
			SupportedTypes: []types.FileType{types.FileTypeCSV},
			ChainConfig:    chainConfig,
			FilenamePrefixPatterns: []string{
				`(?i)^Interspar[_-]?`,
				`(?i)^Spar[_-]?`,
				`(?i)^cjenik[_-]?`,
			},
		},
		ColumnMapping:            intersparColumnMapping,
		AlternativeColumnMapping: &intersparColumnMappingAlt,
	}

	baseAdapter, err := base.NewBaseCsvAdapter(adapterConfig)
	if err != nil {
		return nil, fmt.Errorf("failed to create base CSV adapter: %w", err)
	}

	return &IntersparAdapter{
		BaseCsvAdapter: baseAdapter,
	}, nil
}

// SetDiscoveryDate sets the date to use for discovery filtering
func (a *IntersparAdapter) SetDiscoveryDate(date string) {
	a.discoveryDate = date
}

// Discover discovers available Interspar price files via JSON API
// Interspar provides a JSON API endpoint: /datoteke_cjenici/Cjenik{YYYYMMDD}.json
func (a *IntersparAdapter) Discover(targetDate string) ([]types.DiscoveredFile, error) {
	discoveredFiles := make([]types.DiscoveredFile, 0)

	// Use provided date or default to today
	date := targetDate
	if date == "" {
		date = a.discoveryDate
	}
	if date == "" {
		date = time.Now().Format("2006-01-02")
	}

	// Convert date from YYYY-MM-DD to YYYYMMDD format for the API
	dateForApi := strings.ReplaceAll(date, "-", "")

	// Construct the JSON API URL
	apiUrl := fmt.Sprintf("https://www.spar.hr/datoteke_cjenici/Cjenik%s.json", dateForApi)
	fmt.Printf("[DEBUG] Fetching Interspar JSON API: %s\n", apiUrl)

	resp, err := a.HTTPClient().Get(apiUrl)
	if err != nil {
		fmt.Printf("[ERROR] Failed to fetch Interspar JSON API: %v\n", err)
		return nil, fmt.Errorf("failed to fetch Interspar JSON API: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		fmt.Printf("[ERROR] Interspar JSON API returned status %d\n", resp.StatusCode)
		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
	}

	bodyBytes, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response body: %w", err)
	}

	var data intersparJsonResponse
	if err := json.Unmarshal(bodyBytes, &data); err != nil {
		return nil, fmt.Errorf("failed to parse JSON response: %w", err)
	}

	if len(data.Files) == 0 {
		fmt.Printf("[DEBUG] No files found in JSON response for date %s\n", date)
		return discoveredFiles, nil
	}

	fmt.Printf("[DEBUG] Found %d file(s) in JSON response\n", len(data.Files))

	lastModified, _ := time.Parse("2006-01-02", date)
	for _, file := range data.Files {
		discoveredFiles = append(discoveredFiles, types.DiscoveredFile{
			URL:          file.URL,
			Filename:     file.Name,
			Type:         types.FileTypeCSV,
			Size:         nil,
			LastModified: types.TimePtr(lastModified),
			Metadata: map[string]string{
				"source":       "interspar_json_api",
				"discoveredAt": time.Now().Format(time.RFC3339),
				"portalDate":   date,
				"sha":          file.SHA,
			},
		})
	}

	return discoveredFiles, nil
}

// ExtractStoreIdentifierFromFilename extracts store identifier from Interspar filename
func (a *IntersparAdapter) ExtractStoreIdentifierFromFilename(filename string) string {
	baseName := regexp.MustCompile(`\.(csv|CSV)$`).ReplaceAllString(filename, "")

	// Try to extract 4-digit store code
	match := regexp.MustCompile(`[_-](\d{4})[_-]`).FindStringSubmatch(baseName)
	if len(match) >= 2 {
		return match[1]
	}

	// Try to extract location name after Interspar prefix
	locationMatch := regexp.MustCompile(`(?i)^(?:Interspar|Spar)[_-]?(.+?)(?:[_-]\d{4}[_-]\d{2}[_-]\d{2})?$`).FindStringSubmatch(baseName)
	if len(locationMatch) >= 2 {
		return locationMatch[1]
	}

	// Fallback to base class method
	dummyFile := types.DiscoveredFile{Filename: filename}
	if id := a.ExtractStoreIdentifier(dummyFile); id != nil {
		return id.Value
	}
	return ""
}

// ExtractStoreMetadata extracts store metadata from Interspar filename
// Pattern: {type}_{city}_{address...}_{storeId}_interspar_{city}_{code}_{date}_{time}.csv
func (a *IntersparAdapter) ExtractStoreMetadata(file types.DiscoveredFile) *types.StoreMetadata {
	baseName := regexp.MustCompile(`\.(csv|CSV)$`).ReplaceAllString(file.Filename, "")
	parts := strings.Split(baseName, "_")
	if len(parts) < 8 {
		storeID := a.ExtractStoreIdentifierFromFilename(file.Filename)
		return &types.StoreMetadata{
			Name: fmt.Sprintf("%s %s", a.Name(), storeID),
		}
	}

	storeType := parts[0]
	city := parts[1]

	// Find "interspar" marker
	intersparIdx := -1
	for i, p := range parts {
		if strings.ToLower(p) == "interspar" {
			intersparIdx = i
			break
		}
	}

	if intersparIdx == -1 || intersparIdx < 3 {
		storeID := a.ExtractStoreIdentifierFromFilename(file.Filename)
		return &types.StoreMetadata{
			Name: fmt.Sprintf("%s %s", a.Name(), storeID),
		}
	}

	address := strings.Join(parts[2:intersparIdx-1], " ")

	return &types.StoreMetadata{
		Name:      fmt.Sprintf("Interspar %s", titleCase(city)),
		Address:   titleCase(address),
		City:      titleCase(city),
		StoreType: titleCase(storeType),
	}
}
